File: source/RolXServer/RolXServer/Projects/WebApi/Resource/Activity.cs
// -----------------------------------------------------------------------
// <copyright file="Activity.cs" company="Christian Ewald">
// Copyright (c) Christian Ewald. All rights reserved.
// Licensed under the MIT license.
// See LICENSE.md in the project root for full license information.
// </copyright>
// -----------------------------------------------------------------------

namespace RolXServer.Projects.WebApi.Resource;

/// <summary>
/// An activity in a subproject.
/// </summary>
public sealed record Activity(
    int Id,
    int Number,
    string Name,
    string StartDate,
    string? EndDate,
    int BillabilityId,
    string BillabilityName,
    bool IsBillable,
    long Budget,
    long Planned,
    long Actual,
    string ProjectName,
    string SubprojectName,
    string CustomerName,
    string FullNumber,
    string FullName,
    string AllSubprojectNames);

File: source/rolx-client/src/app/projects/core/activity.ts
import { Duration, TransformAsDuration } from '@app/core/util/duration';
import { TransformAsIsoDate } from '@app/core/util/iso-date';
import { assertDefined } from '@app/core/util/utils';
import * as moment from 'moment';

export class Activity {
  id!: number;
  number!: number;
  name!: string;

  @TransformAsIsoDate()
  startDate!: moment.Moment;

  @TransformAsIsoDate()
  endDate?: moment.Moment;

  billabilityId!: number;
  billabilityName!: string;
  isBillable!: boolean;

  @TransformAsDuration()
  budget!: Duration;

  @TransformAsDuration()
  planned!: Duration;

  @TransformAsDuration()
  actual!: Duration;

  projectName!: string;
  subprojectName!: string;
  customerName!: string;
  fullNumber!: string;
  fullName!: string;
  allSubprojectNames!: string;

  validateModel(): void {
    assertDefined(this, 'id');
    assertDefined(this, 'number');
    assertDefined(this, 'name');
    assertDefined(this, 'startDate');
    assertDefined(this, 'billabilityId');
    assertDefined(this, 'billabilityName');
    assertDefined(this, 'isBillable');
    assertDefined(this, 'budget');
    assertDefined(this, 'planned');
    assertDefined(this, 'actual');
    assertDefined(this, 'projectName');
    assertDefined(this, 'subprojectName');
    assertDefined(this, 'customerName');
    assertDefined(this, 'fullNumber');
    assertDefined(this, 'fullName');
    assertDefined(this, 'allSubprojectNames');
  }

  isOpenAt(date: moment.Moment): boolean {
    return date >= this.startDate && (this.endDate == null || date <= this.endDate);
  }
}
